palace/utils/geodata.cpp

```diff
@@ -1571,7 +1615,344 @@ std::unique_ptr<mfem::Mesh> LoadMesh(const std::string &path, bool remove_curvat
+  if (mesh == orig_mesh.get())
+  {
+    return;
+  }
+
+  // The previous splitting functions remove curvature information from the new mesh. So, if
+  // needed, we interpolate it onto the new mesh with GSLIB. The topology of the mesh must
+  // be finalized for this to work (assumes FinalizeTopology has been called on the new
+  // mesh).
+  if (preserve_curvature && orig_mesh->GetNodes())
+  {
+    // Prepare to interpolate the grid function for high-order nodes from the old mesh to
+    // the new one. This first sets up the new mesh as a linear mesh and constructs a
+    // separate high-order grid function for storing the interpolated nodes.
+    new_mesh.EnsureNodes();
+    const mfem::GridFunction *nodes = orig_mesh->GetNodes();
+    const mfem::FiniteElementSpace *fespace = nodes->FESpace();
+    mfem::Ordering::Type ordering = fespace->GetOrdering();
+    int order = fespace->GetMaxElementOrder();
+    int sdim = orig_mesh->SpaceDimension();
+    bool discont =
+        (dynamic_cast<const mfem::L2_FECollection *>(fespace->FEColl()) != nullptr);
+    mfem::FiniteElementSpace new_fespace(&new_mesh, fespace->FEColl(), sdim, ordering);
+    mfem::GridFunction new_nodes(&new_fespace);
+
+    mfem::Array<int> vdofs;
+    mfem::Vector vals, elem_vals, xyz;
+    mfem::DenseMatrix pointmat;
+    int start = 0;
+    for (int e = 0; e < orig_mesh->GetNE(); e++)
+    {
+      // Get the high-order nodes restricted to this parent element, always returned
+      // byNODES.
+      fespace->GetElementVDofs(e, vdofs);
+      nodes->GetSubVector(vdofs, vals);
+
+      // Find all child elements of this parent in the split mesh and restore their correct
+      // original attribute.
+      int attr = new_mesh.GetAttribute(start);
+      int end = start;
+      while (end < new_mesh.GetNE() && new_mesh.GetAttribute(end) == attr)
+      {
+        new_mesh.SetAttribute(end++, orig_mesh_attr[e]);
+      }
+
+      // Special case: Parent element is unsplit, no interpolation is needed.
+      if (end == start + 1)
+      {
+        new_fespace.GetElementVDofs(start, vdofs);
+        new_nodes.SetSubVector(vdofs, vals);
+        start = end;
+        continue;
+      }
+
+      // Create a list of points at which to interpolate the high order node information.
+      int npts = 0, offset = 0;
+      for (int i = start; i < end; i++)
+      {
+        npts += new_fespace.GetFE(i)->GetNodes().GetNPoints();
+      }
+      xyz.SetSize(npts * sdim);
+      for (int i = start; i < end; i++)
+      {
+        const mfem::FiniteElement &fe = *new_fespace.GetFE(i);
+        mfem::ElementTransformation &T = *new_mesh.GetElementTransformation(i);
+        T.Transform(fe.GetNodes(), pointmat);
+        for (int j = 0; j < pointmat.Width(); j++)
+        {
+          for (int d = 0; d < sdim; d++)
+          {
+            // Use default ordering byNODES.
+            xyz(d * npts + offset + j) = pointmat(d, j);
+          }
+        }
```

Comment:
For this loop, rather than jumping back and forwards by npts in xyz for each write, it might be more cache coherent to have the d loop on the outer. This loops the slow way through pointmat (assuming column major), but pointmat is smaller than xyz, so the npt jumps about in xyz might be causing more reads etc.

Reply:
Yeah good call about `pointmat` being small so the jumps not being a big issue. Fixed.